#106: Dead code warning when setting a default for routes parameters (closes #106)

calippo opened the pull request:
Closes #106

## Test Plan

### tests performed
> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.}

### tests not performed (domain coverage)
> At times not everything can be tested, and writing what hasn't been tested is just as important as writing what has been tested.

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.}

gabro commented on core/src/main/scala/MetaDataMacro.scala:
```diff
@@ -36,9 +36,9 @@ object MetaDataMacro extends MetaDataMacro {
+        }.getOrElse(EmptyTree)
 
-        q"($methodName -> $operationType.map { o => MethodMetaData(o) })"
+        q"($methodName -> Some(MethodMetaData($operationType)))"
```
Instead of doing `getOrElse` and then rewrapping, you can do

```scala
$operationType.map(MethodMetaData(_)).orElse(Some(EmptyTree))
```

or

```scala
val operationType = m.annotations.collectFirst {
  // ...
}.orElse(Some(EmptyTree))

q"($methodName -> $operationType.map(MethodMetadata(_)))"
```

gabro committed:
Add proper fix for unused code warning

weakTypeOf[A].decls returns a list that includes some synthetics methods
that represent the default values of Scala methods.

For example, given

```scala
def getPuppy(token: Auth, wa: Int = 42): Future[Either[Nope, Dog]]
```

`decls` would give

```scala
Scope{
  def getPuppy(token: wiro.Auth,wa: Int):
    scala.concurrent.Future[Either[wiro.apps.controllers.Nope,wiro.apps.models.Dog]];`
  def getPuppy$default$2: Int
    @scala.annotation.unchecked.uncheckedVariance
}
```

The latter needs to be explicitly excluded, or we would generate code
that doesn't get executed. The `.isParamWithDefault` method of
`MethodApi` from scala-reflect provides exactly the API we need.

gabro commented on core/src/main/scala/MetaDataMacro.scala:
```diff
@@ -39,9 +39,9 @@ object MetaDataMacro extends MetaDataMacro {
     }
 
-    q"Map(..$decls) collect { case (k, Some(v)) => (k -> v) }"
+    q"_root_.scala.collection.immutable.Map(..$decls) collect { case (k, Some(v)) => (k -> v) }"
```
this is for hygiene.
